### PR TITLE
feature(slate): Add isElementType utility function to Element Interface

### PR DESCRIPTION
**Pull request by imdbsd**
**Description**
Trying to add isElementType function support like what suggested in the proposal

**Issue**
#4327 

**Example**

https://user-images.githubusercontent.com/16449020/123110300-87840780-d46e-11eb-965e-3e1c9d4666bd.mp4

**Context**
Not sure if this is the correct implementation like what suggested in proposal, since node.type will has type `string` instead of the string literal like implemented in CustomType. I hope there is a suggestion about how to make it become string literal

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [X] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [X] The relevant examples still work. (Run examples with `yarn start`.)
- [X] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

### CODE REVIEW

**Comment by imdbsd** [comment]
Also, because im share the isElement type function to it, im kinda confuse where to put the function. Under interface/shared.tsx maybe?

File: packages/slate/src/interfaces/element.ts [hunk 1]
```diff
@@ -18,9 +18,24 @@ export interface ElementInterface {
   isElement: (value: any) => value is Element
   isElementList: (value: any) => value is Element[]
   isElementProps: (props: any) => props is Partial<Element>
+  isElementType: (
+    value: any,
+    type: string
+  ) => value is Element & { type: string }
```
Comment (BrentFarese) [inline packages/slate/src/interfaces/element.ts:24]: I would suggest maybe using a Generic type for this:

```suggestion
  isElementType: <T extends Element>(
    value: any,
    discriminantVal: string,
    discriminantKey?: string
  ) => value is T
```

File: packages/slate/src/interfaces/element.ts [hunk 2]
```diff
@@ -58,6 +66,20 @@ export const Element: ElementInterface = {
...
+
+  isElementType: (
+    value: any,
+    type: string
+  ): value is Element & { type: string } => {
+    if (value?.type === type && isElement(value)) {
+      return true
+    }
+    return false
+  },
```
Comment (BrentFarese) [inline packages/slate/src/interfaces/element.ts:81]: Maybe do something like the below to make the func more flexible?

```suggestion
  isElementType: (
    value: any,
    discriminantVal: string,
    discriminantKey: string = 'type'
  ): value is Element & { type: string } => {
    return isElement(value) && value[discriminantKey] === discriminantVal;
  },
```

Comment (BrentFarese) [reply packages/slate/src/interfaces/element.ts:81]: The point being that not everyone uses `type` (although most people do) so maybe we make the func flexible to other [discriminants](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions).

Comment (imdbsd) [reply packages/slate/src/interfaces/element.ts]: haaa i see, now im understand why `type` is not in Element type

File: packages/slate/src/interfaces/element.ts (hunk 1 above)
Comment (imdbsd) [reply packages/slate/src/interfaces/element.ts]: implemented, and also im adding default Extend value to `Element & {type: string}`

**Comment by imdbsd** [comment]
Hi @BrentFarese thank you for reviewing this PR, im already implemented your suggestion. Please review it again.

**Comment by imdbsd** [comment]
> Do you need to add `{ type: string }` here?

File: packages/slate/test/interfaces/Element/isElement/elementType.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+export const input = {
+  type: 'paragraph',
+  children: [{ text: '' }],
+}
+export const test = value => {
+  return Element.isElementType(value, 'heading-large')
+}
+
+export const output = true
```
Comment (BrentFarese) [inline packages/slate/test/interfaces/Element/isElement/elementType.tsx:11]: Shouldn't this return `false` since `'heading-large'` is not `'paragraph`'?

Comment (BrentFarese) [reply packages/slate/test/interfaces/Element/isElement/elementType.tsx:11]: We should probably add tests for `true` and `false` cases.

**Comment by BrentFarese** [comment]
> ~im thinking about using `isElementType` without the discriminant key and the user doesn't need to add the generic for accessing the `element.type`.~
> since the purpose is to safely access `element.type` as default key, im adding `{ type: string}` as the default extends, do you have any recomendation or suggestion ?

I like what you did in your recent commit by making it the default type, but it's not required to have a `type` key.

File: packages/slate/test/interfaces/Element/isElement/elementType.tsx (hunk 3 above)
Comment (imdbsd) [reply packages/slate/test/interfaces/Element/isElement/elementType.tsx]: My bad. Updated and adding some test for the discriminant key usage

**Review by BrentFarese — APPROVED** [review]
Looks good to me.

**Comment by horacioh** [comment]
do we need to add documentation for this method right? cc @imdbsd

**Pull request merged**